### PR TITLE
Upgrade to Micronaut 4

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

**Review by wetted — CHANGES_REQUESTED** [review]
The top-level `build.gradle` needs the following for the `docs` task.

```
repositories {
    mavenCentral()
    maven { url "https://s01.oss.sonatype.org/content/repositories/snapshots/" }
}
```

I noted a couple nit-picky things. Otherwise, looks good to me.

File: buildSrc/src/main/groovy/io.micronaut.build.internal.r2dbc-dependencies.gradle [hunk 1]
```diff
@@ -0,0 +1,15 @@
+repositories {
+    mavenCentral()
+    maven { url "https://s01.oss.sonatype.org/content/repositories/snapshots/" }
+}
+configurations {
+    all {
+        resolutionStrategy {
+            dependencySubstitution {
+                substitute(module("org.codehaus.groovy:groovy"))
+                        .using(module("org.apache.groovy:groovy:4.0.5"))
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply buildSrc/src/main/groovy/io.micronaut.build.internal.r2dbc-dependencies.gradle:10]: it should be `4.0.7` now.

File: buildSrc/src/main/groovy/io.micronaut.build.internal.r2dbc-module.gradle [hunk 2]
```diff
@@ -0,0 +1,12 @@
+plugins {
+    id "java-library"
+    id "io.micronaut.build.internal.r2dbc-dependencies"
+    id "io.micronaut.build.internal.module"
+}
+
+micronautBuild {
+    binaryCompatibility {
+        // Enable before Micronaut 4 release
+        enabled.set(false)
```
Comment (sdelamo) [inline buildSrc/src/main/groovy/io.micronaut.build.internal.r2dbc-module.gradle:10]: why do we need this ? Are not all the modules already released? Can we remove this?

File: doc-examples/example-java/src/main/java/example/Application.java [hunk 3]
```diff
@@ -1,36 +1,7 @@
 package example;
 
-import com.mysql.cj.exceptions.*;
-import io.micronaut.core.annotation.TypeHint;
 import io.micronaut.runtime.Micronaut;
 
-@TypeHint({
-        AssertionFailedException.class,
```
Comment (sdelamo) [inline doc-examples/example-java/src/main/java/example/Application.java:8]: Are these `@TypeHint` clases no longer needed?

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -1,25 +1,52 @@
...
+micronaut-serde = "2.0.0-SNAPSHOT"
+micronaut-test = "4.0.0-SNAPSHOT"
+micronaut-kotlin = "4.0.0-SNAPSHOT"
+micronaut-data = "4.0.0-SNAPSHOT"
+micronaut-sql = "5.0.0-SNAPSHOT"
+micronaut-flyway = "6.0.0-SNAPSHOT"
 
-h2 = "1.4.200"
-ojdbc11 = "21.7.0.0"
+groovy = "4.0.6"
```
Comment (sdelamo) [inline gradle/libs.versions.toml:14]:
```suggestion
groovy = "4.0.7"
```

File: settings.gradle [hunk 5]
```diff
@@ -1,36 +1,26 @@
...
 dependencyResolutionManagement {
     repositories {
         mavenCentral()
+        maven { url "https://s01.oss.sonatype.org/content/repositories/snapshots/" }
     }
 }
 
 micronautBuild {
     importMicronautCatalog()
+    importMicronautCatalog("micronaut-reactor")
```
Comment (sdelamo) [inline settings.gradle:16]: add snapshot

File: test-graalvm/common/src/main/java/testgraalvm/controllers/dto/OwnerDto.java [hunk 6]
```diff
@@ -15,15 +15,24 @@
...
 public class OwnerDto {
 
     private Long id;
     private String name;
     private int age;
 
+    public OwnerDto() {
+    }
+
+    public OwnerDto(Long id, String name, int age) {
```
Comment (sdelamo) [inline test-graalvm/common/src/main/java/testgraalvm/controllers/dto/OwnerDto.java:30]: should this constructor be annotated with `@Creator`?

File: test-graalvm/common/src/main/java/testgraalvm/domain/Owner.java [hunk 7]
```diff
@@ -29,6 +29,20 @@ public class Owner {
...
 
+    public Owner() {
+    }
+
+    public Owner(String name, int age) {
+        this.name = name;
+        this.age = age;
+    }
+
+    public Owner(Long id, String name, int age) {
```
Comment (sdelamo) [inline test-graalvm/common/src/main/java/testgraalvm/domain/Owner.java:40]: should this contractor be annotated with `@Creator`?

File: test-graalvm/common/src/main/java/testgraalvm/repositories/OwnerRepository.java [hunk 8]
```diff
@@ -19,7 +19,7 @@ public interface OwnerRepository extends ReactorCrudRepository<Owner, Long> {
 
     Mono<OwnerDto> findByName(String name);
 
-    Mono<Owner> getByName(String name);
+    Mono<OwnerDto> getByName(String name);
 
-    <S extends Owner> Publisher<S> saveAll(@Valid @NotNull Iterable<S> entities, Connection connection);
+    Flux<Owner> saveAll(@Valid @NotNull Iterable<Owner> entities, Connection connection);
```
Comment (sdelamo) [inline test-graalvm/common/src/main/java/testgraalvm/repositories/OwnerRepository.java:24]: Why are we changing this from Publisher to Flux?

File: test-graalvm/common/src/main/java/testgraalvm/repositories/PetRepository.java [hunk 9]
```diff
@@ -19,5 +19,5 @@ public interface PetRepository extends ReactorCrudRepository<Pet, Long> {
 
     Mono<PetDto> findByName(String name);
 
-    <S extends Pet> Publisher<S> saveAll(@Valid @NotNull Iterable<S> entities, Connection connection);
+    Flux<Pet> saveAll(@Valid @NotNull Iterable<Pet> entities, Connection connection);
```
Comment (sdelamo) [inline test-graalvm/common/src/main/java/testgraalvm/repositories/PetRepository.java:22]: Why are we changing this from Publisher to Flux?

File: test-graalvm/oracle/build.gradle [hunk 10]
```diff
@@ -0,0 +1,9 @@
+plugins {
+    id 'io.micronaut.build.internal.r2dbc-testproject'
+}
+
+dependencies {
+    runtimeOnly(libs.managed.r2dbc.oracle)
+    runtimeOnly("com.oracle.database.jdbc:ojdbc11:21.8.0.0")
```
Comment (sdelamo) [inline test-graalvm/oracle/build.gradle:7]: can this version be in libs.versions.toml?

File: r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java [hunk 11]
```diff
@@ -35,15 +35,14 @@ public void beforeAnalysis(BeforeAnalysisAccess access) {
...
-
             AutomaticFeatureUtils.initializeAtRunTime(
                     access,
                     "org.mariadb.r2dbc.MariadbConnectionFactoryProvider",
                         "org.mariadb.r2dbc.client.ClientBase",
                         "org.mariadb.r2dbc.message.flow.AuthenticationFlow",
                         "org.mariadb.r2dbc.util.DefaultHostnameVerifier"
                     );
-
+            AutomaticFeatureUtils.addResourcePatterns("project.properties");
```
Comment (sdelamo) [inline r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java:45]: can you explain this line?

File: buildSrc/build.gradle [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    gradlePluginPortal()
+    mavenCentral()
+}
+
+dependencies {
+    implementation libs.gradle.micronaut
+    implementation libs.gradle.kotlin
+    implementation libs.gradle.kotlin.allopen
+    implementation libs.gradle.kotlin.noarg
+//    implementation libs.gradle.test.logger
```
Comment (sdelamo) [inline buildSrc/build.gradle:15]:
```suggestion
```

File: .github/workflows/corretto.yml [hunk 13]
```diff
@@ -0,0 +1,36 @@
+name: Corretto CI
```
Comment (sdelamo) [inline .github/workflows/corretto.yml:1]: can you explain why we add a `corretto.yml` file?

**Comment by sdelamo** [comment]
Is this a duplicated of https://github.com/micronaut-projects/micronaut-r2dbc/pull/283?

File: .github/workflows/corretto.yml (hunk 13 above)
Comment (dstepanov) [reply .github/workflows/corretto.yml:1]: I just copied the latest github workflows from Corr

File: r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java (hunk 11 above)
Comment (dstepanov) [reply r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java:45]: What do you want to explain? The change is needed for the native execution

File: test-graalvm/oracle/build.gradle (hunk 10 above)
Comment (dstepanov) [reply test-graalvm/oracle/build.gradle:7]: After sql module is updated

File: test-graalvm/common/src/main/java/testgraalvm/repositories/PetRepository.java (hunk 9 above)
Comment (dstepanov) [reply test-graalvm/common/src/main/java/testgraalvm/repositories/PetRepository.java:22]: There was an attempt to fix the tests, but this is not needed

File: test-graalvm/common/src/main/java/testgraalvm/controllers/dto/OwnerDto.java (hunk 6 above)
Comment (dstepanov) [reply test-graalvm/common/src/main/java/testgraalvm/controllers/dto/OwnerDto.java:30]: Not needed

File: test-graalvm/common/src/main/java/testgraalvm/domain/Owner.java (hunk 7 above)
Comment (dstepanov) [reply test-graalvm/common/src/main/java/testgraalvm/domain/Owner.java:40]: Not needed

File: doc-examples/example-java/src/main/java/example/Application.java (hunk 3 above)
Comment (dstepanov) [reply doc-examples/example-java/src/main/java/example/Application.java:8]: There shouldn’t have been there in a first place

File: buildSrc/src/main/groovy/io.micronaut.build.internal.r2dbc-module.gradle (hunk 2 above)
Comment (dstepanov) [reply buildSrc/src/main/groovy/io.micronaut.build.internal.r2dbc-module.gradle:10]: Binary compatibility is not needed to be tracked across the major modules

**Review by graemerocher — APPROVED** [review]

File: .github/workflows/corretto.yml (hunk 13 above)
Comment (sdelamo) [reply .github/workflows/corretto.yml]: We don't have a corretto.yml file in the project template https://github.com/micronaut-projects/micronaut-project-template/tree/master/.github/workflows ?

Are we testing against corretto.yml  in the R2DBC project for any specific reason?

File: r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java (hunk 11 above)
Comment (sdelamo) [reply r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java:45]: is `project.properties` a file specific for R2DBC projects?

File: .github/workflows/corretto.yml (hunk 13 above)
Comment (dstepanov) [reply .github/workflows/corretto.yml]: No, it was a mistake

File: r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java (hunk 11 above)
Comment (dstepanov) [reply r2dbc-core/src/main/java/io/micronaut/r2dbc/graalvm/R2dbcAutomaticFeature.java:45]: No, It's part of the internal driver impl

**Pull request merged**